### ISSUE TITLE
CPU memory leak when using dynamo

### ISSUE THREAD

**Issue by pratogab** [huggingface/accelerate#1987]
### System Info

```Shell
accelerate v0.22
torch v2.0
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Running my code with the following accelerate config:
```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
no issues, but then if I use torch dynamo:
```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
dynamo_config:
  dynamo_backend: INDUCTOR
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
I get a CPU RAM memory leak. This might be reproducible via the accelerate examples such as `nlp_example.py`

### Expected behavior

There shouldn't be any memory leaks.

**Comment by BenjaminBossan** [comment]
Thanks for reporting. I tried your two configs on my local machine and with the `nlp_example.py` and could not observe any memory leak. Both with and without dynamo backend, the memory usage was constant during training, so there was no leakage. On top of that, the total memory usage was also very similar between the two settings.

Could you please provide more details, e.g. total memory usage and how it increases over time (e.g. at the end of each batch? each epoch?) and by how much?

**Comment by pratogab** [comment]
Minimal reproducer:
```yaml
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
dynamo_config:
  dynamo_backend: INDUCTOR
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

```python
from accelerate import Accelerator
from transformers import AutoTokenizer, AutoModelForCausalLM


def main():
    accelerator = Accelerator()

    device = accelerator.device

    model = AutoModelForCausalLM.from_pretrained('facebook/opt-125m').to(device)
    tokenizer = AutoTokenizer.from_pretrained('facebook/opt-125m')

    model = accelerator.prepare(model)

    while True:
        x = tokenizer(['I like turtles.'], return_tensors='pt')
        x = x.to(device)
        output = model(**x)

if __name__ == '__main__':
    main()
```

`env OMP_NUM_THREADS=4 TOKENIZERS_PARALLELISM=false accelerate launch --config_file=<accelerate_config_path> ./example_script.py`

**Comment by BenjaminBossan** [comment]
Thanks for providing the script and config, I ran this locally and could not reproduce any memory leak. Here is a screenshot from my nvtop:

![Screenshot from 2023-09-28 14-23-04](https://github.com/huggingface/accelerate/assets/6229650/5b401bc7-7f92-40c4-a586-2086946c77b6)

(note that the blip comes from another process)

I also tried on a T4, with the same result. In both cases, I this ran for several minutes without any increase in memory usage. I also checked main RAM and it was constant. So it is probably something that happens only in very specific circumstances. If you could try to reproduce the behavior on a different machine, it would be helpful.

**Comment by pratogab** [comment]
The memory leak is for CPU memory, not GPU memory.

Here's the node info:

> Distributor ID: Ubuntu
Description:    Ubuntu 18.04.6 LTS
Release:        18.04
Codename:       bionic

> Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  1
Core(s) per socket:  20
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
Stepping:            7
CPU MHz:             803.126
CPU max MHz:         3900.0000
CPU min MHz:         800.0000
BogoMIPS:            4200.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            28160K
NUMA node0 CPU(s):   0-19
NUMA node1 CPU(s):   20-39
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_pkg_req pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

> +---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.10              Driver Version: 535.86.10    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-SXM2-32GB           On  | 00000000:15:00.0 Off |                    0 |
| N/A   28C    P0              40W / 300W |      3MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

package versions:
> accelerate 0.22.0
cuda-runtime              11.8.0                   0         nvidia
cuda-version              11.8                 h70ddcb2_2    conda-forge
cudatoolkit               11.8.0              h4ba93d1_12    conda-forge
cudnn                     8.8.0.121            h838ba91_3    conda-forge
python                    3.10.12         hd12c33a_0_cpython    conda-forge
pytorch                   2.0.1           py3.10_cuda11.8_cudnn8.7.0_0    pytorch
pytorch-cuda              11.8                 h7e8668a_5    pytorch
transformers              4.33.1                   pypi_0    pypi

Here's the CPU memory every 60 seconds **with torch dynamo**:
```
while true; do ps -u pratogab --no-headers -o rss | awk '{sum+=$1} END {print sum/1024 " MB"}'; sleep 60; done
14194.2 MB
14337.6 MB
14480 MB
14623 MB
14764.6 MB
14906.5 MB
15045.9 MB
15188.6 MB
15330.1 MB
```

And **without torch dynamo**:
```
while true; do ps -u pratogab --no-headers -o rss | awk '{sum+=$1} END {print sum/1024 " MB"}'; sleep 60; done
1019.81 MB
1020.14 MB
1020.21 MB
1020.13 MB
1020.04 MB
1020.23 MB
1019.98 MB
1020.05 MB
1020.02 MB
1020.03 MB
1020.18 MB
```

If you want me to reproduce this on other machines, I can, happened on A100 nodes as well.

**Comment by BenjaminBossan** [comment]
Thanks for providing more info, I somehow assumed that you were talking about GPU memory.

I re-ran the script and monitored main memory, comparing accelerate with dynamo, without dynamo, and pure PyTorch with dynamo. Here are the results (measured at 10 sec interval):

![image](https://github.com/huggingface/accelerate/assets/6229650/c467b16e-c9f5-479f-bac5-39cf624fd934)

So without dynamo, memory usage is pretty much constant but with dynamo, it slowly increases over time, indicating a memory leak. Interestingly, the increase seems to be pretty similar between accelerate vs pure PyTorch. This makes me think that accelerate is not at fault here.

@pratogab Are you able to replicate the leak using pure PyTorch? For reference, this is the script that I used:

import torch
from transformers import AutoTokenizer, AutoModelForCausalLM


```python
def main():
    model = AutoModelForCausalLM.from_pretrained('facebook/opt-125m').to("cuda")
    tokenizer = AutoTokenizer.from_pretrained('facebook/opt-125m')

    model = torch.compile(model, backend="inductor")

    while True:
        x = tokenizer(['I like turtles.'], return_tensors='pt')
        x = x.to("cuda")
        output = model(**x)

if __name__ == '__main__':
    main()

```

**Comment by pratogab** [comment]
Happens as well without accelerate, just pytorch. I'll open an issue on their github. Sorry for that and thanks.

**Comment by pratogab** [comment]
Issue has been fixed in torch v2.1
